### ISSUE TITLE
Improve app restart handling (after configuration or object field related schema changes)

### ISSUE THREAD

**Issue by mgruner** [zammad/zammad#5377]
### Used Zammad Version

6.3

### Environment

- Installation method: all
- Operating system (if you're unsure: `cat /etc/os-release` ): all
- Database + version:  all
- Elasticsearch version: irrelevant

### Actual behaviour

When making changes to core objects like adding new Ticket fields, schema changes in the database have to be executed. Also, configuration changes like FQDN or HTTP type would require a restart. After that, a restart of all application server processes is needed. This can be automated on package and source installations (via APP_RESTART_CMD), but does not work in containerized environments (hacks are possible for docker, but go against security considerations).

Additionally, for the manual restarts, there is a time of inconsistency when the schema was already changed, but processes like scheduler not yet restarted - which can lead to inconsistent behaviour.

### Expected behaviour

Zammad should be changed in a way that these restarts are either not needed, or work correctly on any installation method in an automated way, reducing times of inconsistent code state to a minimum.

### Steps to reproduce the behaviour

- Install Zammad on kubernetes
- Create a new ticket field and apply the database changes
- Now you need to redeploy the app on Kubernetes manually

### I'm sure this is a bug and no feature request or a general question.

yes

**Comment by mgruner** [comment]
## Ideas

We should evaluate first if the restarts are actually required, or if on the server a refresh of the model objects (`reset_column_information`) would be sufficient, like we do it in the migrations.

### If Restart is Required

We could consider informing all processes about a required change (e.g. through a hidden setting with a number that is incremented). All processes will detect such a change and respond to it by terminating. It's the responsibility of the process manager (systemd, docker compose, kubernetes) to bring up new processes instead.

### If Reload is Enough

All processes are notified about the change (see above), but they don't terminate but instead only refresh their model data and other relevant parts of the API on the back end side. The new mobile and desktop front ends don't need a restart anyway, legacy app still needs a browser reload.

**Comment by dominikklein** [comment]
> The new mobile and desktop front ends don't need a restart anyway

Why do you think the new stack needs no reload?

**Comment by mgruner** [comment]
I was under the impression, because our graphql schema is static and does not change based on the object fields. Is that not correct?

**Comment by dominikklein** [comment]
No, because form fields are build based on the object attributes which are returned from the server. And the attributes are fetched once. So after a new attribute was added/changed it's no longer up to date from this perspecctive.
For sure we could add some subscription for that, but at least it's currently not implemented.

**Comment by mgruner** [comment]
Ok, but that would work without a full app reload, hopefully.

**Comment by mgruner** [comment]
Action cable initializer relies on configured FQDN and HTTP type. It might not work without a reload.

**Comment by mgruner** [comment]
@YetAnotherGerrit we identified another new issue [coordination-scrum/issues/214](https://github.com/zammad/coordination-scrum/issues/214) which should also be addressed as part of this story. It's about safe shutdown handling in the background worker. We added it as a sub-issue.

**Comment by mgruner** [comment]
### Result of https://github.com/zammad/coordination-scrum/issues/192 investigation

- Please see the sub-issue for technical details.
- We found that a restart is not technically required. It would be possible to implement process refresh hooks at runtime. However, this comes with increased complexity and less transparency. Depending on configuration, puma and the scheduler process may fork out sub-processes which must be independently refreshed.
- Therefore we suggest to go with the simpler approach, which is a controlled process self-shutdown (or restart if feasible). This has less complexity and is also better transparent for the admin.

An implementation could include these points:
- When `fqdn` or `http_type` settings are changed, or object attribute migrations are executed, then a new mechanism will signal to all processes in the application cluster that they need to be restarted.
  - This could happen via a new config setting which holds a timestamp. If processes find the timestamp changed, they have to be restarted.
- Processes will react to this signal (e.g. via a maintenance `Thread` in the main process) and shut themselves down in a controlled, graceful way.
  - In the case of puma, there is a possibility to use its [hot-reload mechanism](https://github.com/puma/puma/blob/master/docs/restart.md). But we are not sure if this will play well with process managers like systemd, docker or kubernetes, or if it would be better to do a clean shutdown instead. This needs to be analyzed.
- The process manager is responsible to bring up new instances of the processes automatically (by way of its restart policy). 
- It should be possible to opt out of the auto-shutdown feature via new config setting for special use cases.

We believe that an estimation between M and L might be appropriate here.

**Comment by YetAnotherGerrit** [comment]
Is the list of sub tasks complete, therefore will this story be fulfilled when they are completed?

**Comment by mgruner** [comment]
@YetAnotherGerrit yes, it will be complete.
